### PR TITLE
Refactor to_chars_result to reflect 22.13.2

### PR DESCRIPTION
**Pull request by mborland**
Uses `std::errc` instead of `int`. Also adds definition of `operator==` since defaulting the operator is not available until C++20.

### CODE REVIEW

**Comment by pdimov** [comment]
I used `int` to avoid including the entire `<system_error>` for just `std::errc`. It's a relatively heavy include because it pulls `<string>` and most people don't want that; for embedded uses, `<string>` might even be forbidden.

File: include/boost/charconv/to_chars.hpp [hunk 1]
```diff
@@ -1,24 +1,36 @@
-#ifndef BOOST_CHARCONV_TO_CHARS_HPP_INCLUDED
-#define BOOST_CHARCONV_TO_CHARS_HPP_INCLUDED
-
 // Copyright 2022 Peter Dimov
+// Copyright 2023 Matt Borland
 // Distributed under the Boost Software License, Version 1.0.
 // https://www.boost.org/LICENSE_1_0.txt
```
Comment (pdimov) [inline include/boost/charconv/to_chars.hpp:4]: I deliberately leave the copyright clause inside the `ifndef` to avoid it being scanned, although that's not particularly relevant in 2023.

File: include/boost/charconv/to_chars.hpp [hunk 2]
```diff
@@ -1,24 +1,36 @@
...
 
+#ifndef BOOST_CHARCONV_TO_CHARS_HPP_INCLUDED
+#define BOOST_CHARCONV_TO_CHARS_HPP_INCLUDED
+
+#include <system_error>
```
Comment (pdimov) [inline include/boost/charconv/to_chars.hpp]: As a rule, includes should be ordered such that more general includes follow less general ones. Including <system_error> before <boost/charconv/...> would hide a failure to include <system_error> in <boost/charconv/..> if that latter header uses something from <system_error>.

File: include/boost/charconv/to_chars.hpp [hunk 3]
```diff
@@ -1,24 +1,36 @@
...
 {
-    char * ptr;
-    int ec;
+    char* ptr;
+    std::errc ec;
+    friend bool operator==(const to_chars_result&, const to_chars_result&);
 };
 
-BOOST_CHARCONV_DECL to_chars_result to_chars( char* first, char* last, int value, int base = 10 );
+bool operator==(const to_chars_result& lhs, const to_chars_result& rhs)
```
Comment (pdimov) [inline include/boost/charconv/to_chars.hpp]: inline is missing here.

File: include/boost/charconv/to_chars.hpp [hunk 4]
```diff
@@ -1,24 +1,36 @@
...
+{
+    if (lhs.ptr == rhs.ptr && lhs.ec == rhs.ec)
```
Comment (pdimov) [inline include/boost/charconv/to_chars.hpp]: Why not just `return lhs.ptr == rhs.ptr && lhs.ec == rhs.ec;` here?

File: test/to_chars.cpp [hunk 5]
```diff
@@ -13,7 +13,8 @@ int main()
     int v = 1048576;
     auto r = boost::charconv::to_chars( buffer, buffer + sizeof( buffer ) - 1, v );
 
-    BOOST_TEST_EQ( r.ec, 0 ) && BOOST_TEST_CSTR_EQ( buffer, "1048576" );
+    BOOST_TEST(!std::make_error_code(r.ec));
```
Comment (pdimov) [inline test/to_chars.cpp]: Note that this changes the behavior of always executing the second test even if the previous one fails. This is probably OK here, as the final zero in `buffer` should always be present.

**Comment by mborland** [comment]
> 

That makes sense to me. Are we planning for embedded uses, and need to avoid throwing as well?

**Comment by pdimov** [comment]
In general, each new functionality should come with a test. In this case, we need a test for chars_format and its operator==.

Also, I think that `== default` in C++20 also causes operator!= to be defined, so we need this and a test for it as well.

**Comment by pdimov** [comment]
> Are we planning for embedded uses, and need to avoid throwing as well?

This functionality (directly or indirectly via Boost.JSON) is definitely useful for embedded, yes.

It would definitely be good if these functions can be `noexcept`. I know that some corner floating-point cases require multiprecision arithmetic, so we might need to allocate and hence throw, but it would be good to avoid that in as many cases as possible.

**Comment by pdimov** [comment]
> Also, I think that `== default` in C++20 also causes operator!= to be defined

Indeed it does, https://godbolt.org/z/Yj3YWTGzc.

**Comment by pdimov** [comment]
Wait, I used `chars_format` instead of `to_chars_result` by mistake. It's https://godbolt.org/z/zx5PGnvro, and the op!= is synthesized by C++20.

**Comment by pdimov** [comment]
A general comment on the inline friend in the standard: the function is declared as a `friend` inside the struct definition, even though all the members are public and therefore there's no need for it being a friend at all, to denote an inline friend. That is, a function whose definition is inline in the friend declaration.

This is a special kind of function called "inline friend" or "hidden friend" because it's not visible to normal lookup, and therefore does not contribute to the enormous overload sets that are usually formed when code does `x == y` where one of `x` or `y` is in namespace `std`. A hidden friend only participates in overload resolution when the class is mentioned.

TL;DR:

```
struct to_chars_result
{
    char* ptr;
    int ec;

    friend bool operator==(const to_chars_result& r1, const to_chars_result& r2)
    {
        return r1.ptr == r2.ptr && r1.ec == r2.ec;
    }
};
```
